File: crash_course/ch03/gofiles/logger3.go
package main

import "fmt"

type Logger interface {
	Log(from, to uint64, amount float64)
}

type ConsoleLogger struct{}

func (c *ConsoleLogger) Log(from, to uint64, amount float64) {
	fmt.Printf("[Cons] %d -> %d: %.6f\n", from, to, amount)
}

type FileLogger struct{}

func (f *FileLogger) Log(from, to uint64, amount float64) {
	fmt.Printf("[File] %d -> %d: %.6f\n", from, to, amount)
}

type Bank struct {
	Logger
}

func NewBank() *Bank {
	return &Bank{Logger: &ConsoleLogger{}}
}

func (b *Bank) SetLogger(logger Logger) {
	b.Logger = logger
}

func (b *Bank) MakeTransfer(from, to uint64, amount float64) {
	b.Log(from, to, amount)
}

File: crash_course/ch03/logger3.cpp

// A complete C++ Program
#include <iostream>
#include <iomanip>

struct Logger
{
    virtual void log(long from, long to, double amount) const = 0;
    virtual ~Logger() = default;
};

struct FileLogger : Logger
{
    void log(long from, long to, double amount) const override
    {
        std::cout << "[File] " << from << " -> " << to << ": "
            << std::setw(5) << std::fixed << amount << "\n";
    }
};

struct ConsoleLogger : Logger
{
    void log(long from, long to, double amount) const override
    {
        std::cout << "[Cons] " << from << " -> " << to << ": "
            << std::setw(5) << std::fixed << amount << "\n";
    }
};

struct Bank
{
    Bank(): m_logger{new ConsoleLogger{}}
    {}

    void set_logger(Logger* logger)
    {
        if (m_logger)
        {
            delete m_logger;
        }
        m_logger = logger;
    }

    void make_transfer(long from, long to, double amount)
    {
        if (m_logger)
        {
            m_logger->log(from, to, amount);
        }
    }

private:
    Logger* m_logger = nullptr;
};


int main()
{
    Bank bank;
    bank.make_transfer(1000, 2000, 49.95);
    bank.make_transfer(2000, 4000, 20.00);
    bank.set_logger(new FileLogger{});
    bank.make_transfer(3000, 2000, 75.00);

    return 0;
}
